RecoLocalCalo/HGCalRecProducers/plugins/HGCalCLUEAlgo.cc

```diff
@@ -216,33 +220,104 @@ math::XYZPoint HGCalCLUEAlgo::calculatePosition(const std::vector<int>& v, const
-      for (int yBin = search_box[2]; yBin < search_box[3] + 1; ++yBin) {
-        int binId = lt.getGlobalBinByBin(xBin, yBin);
-        size_t binSize = lt[binId].size();
-
-        for (unsigned int j = 0; j < binSize; j++) {
-          unsigned int otherId = lt[binId][j];
-          if (distance(i, otherId, layerId) < delta_c) {
-            cellsOnLayer.rho[i] += (i == otherId ? 1.f : 0.5f) * cellsOnLayer.weight[otherId];
+    float delta(0);
+    if (cellsOnLayer.isSi[i]) {
+      delta = delta_c;
+      std::array<int, 4> search_box = lt.searchBox(
+          cellsOnLayer.x[i] - delta, cellsOnLayer.x[i] + delta, cellsOnLayer.y[i] - delta, cellsOnLayer.y[i] + delta);
+
+      for (int xBin = search_box[0]; xBin < search_box[1] + 1; ++xBin) {
+        for (int yBin = search_box[2]; yBin < search_box[3] + 1; ++yBin) {
+          int binId = lt.getGlobalBinByBin(xBin, yBin);
+          size_t binSize = lt[binId].size();
+
+          for (unsigned int j = 0; j < binSize; j++) {
+            unsigned int otherId = lt[binId][j];
+            if (!cellsOnLayer.isSi[otherId])
+              continue;  //cells in the silicon cannot talk to cells in scintillator
+            if (distance(i, otherId, layerId, false) < delta) {
+              cellsOnLayer.rho[i] += (i == otherId ? 1.f : 0.5f) * cellsOnLayer.weight[otherId];
+            }
+          }
+        }
+      }
+    } else {
+      delta = delta_r;
+      std::array<int, 4> search_box = lt.searchBoxEtaPhi(cellsOnLayer.eta[i] - delta,
+                                                         cellsOnLayer.eta[i] + delta,
+                                                         cellsOnLayer.phi[i] - delta,
+                                                         cellsOnLayer.phi[i] + delta);
+      cellsOnLayer.rho[i] += cellsOnLayer.weight[i];
+      float northeast(0), northwest(0), southeast(0), southwest(0), all(0);
+
+      for (int etaBin = search_box[0]; etaBin < search_box[1] + 1; ++etaBin) {
+        for (int phiBin = search_box[2]; phiBin < search_box[3] + 1; ++phiBin) {
+          int binId = lt.getGlobalBinByBinEtaPhi(etaBin, phiBin);
+          size_t binSize = lt[binId].size();
+
+          for (unsigned int j = 0; j < binSize; j++) {
+            unsigned int otherId = lt[binId][j];
+            if (cellsOnLayer.isSi[otherId])
+              continue;  //cells in the silicon cannot talk to cells in scintillator
+            if (distance(i, otherId, layerId, true) < delta) {
+              // need to round these numbers to 5 decimals to be able to compare floats
+              constexpr float roundDec = 100000.;
+              float iPhi = round(cellsOnLayer.phi[i] * roundDec) / roundDec;
+              float otherphi = cellsOnLayer.phi[otherId];
+              otherphi += (otherphi * iPhi >= 0 || abs(iPhi) < 1.) ? 0. : iPhi > 0. ? 2 * M_PI : -2 * M_PI;
+              float otherPhi = round(otherphi * roundDec) / roundDec;
```

Comment:
Why are we using round? 
The tiles in x, y are done rounding down to the integer, not to the nearest one, maybe we should use a uniform way of tiling.

Reply:
I think the tiling for the scintillator section is the same as the silicon section. The only reason I do the rounding is that without rounding, the comparison logic below for constructing a 2x2 window around a cell will not work. 
For example, I have a cell which has eta-phi values of **1.67560517787933349609375** and 0.55632364749908447265625, and the cell above it in eta-phi space has eta-phi **1.67560505867004394531250** and 0.57814025878906250000000, so I can't put a "==" between the two eta values.

Reply:
Jingyu, a float in C++ is a single precision 32-bit floating point number following the IEEE-754 standard:
[https://en.wikipedia.org/wiki/Single-precision_floating-point_format](https://en.wikipedia.org/wiki/Single-precision_floating-point_format
)

Which typically has 7 significant digits.

I'm not sure what other language you are using for parsing the eta-phi values, but for sure 1.67560517787933349609375 cannot be represented with full precision with a float.

Reply:
Thanks, Felice. But I was a little bit confused. Did you mean that eta=1.67560517787933349609375 should not happen since I specify it to be a float?

Reply:
And I just double-checked, I can confirm that when I print out the eta values of two different cells in the same module (HGCalCLUEAlgo.cc), they both equal **1.67561**, but neither eta1==eta2 nor eta1-eta2==0.f passed. This should be because in better accuracy, one is equal to **1.67560517787933349609375** and other is equal to **1.67560505867004394531250**. The values with better accuracy are printed out by 
``std::cout << std::fixed << std::setprecision(24) ``. 
Let me know if this sounds correct to you. If so, I guess the rounding is needed?
Thanks